### PR TITLE
Falling furniture and items

### PR DESCRIPTION
**Pull request by Coolthulhu**
First point from #6823:

> Unsupported items, furniture, vehicles and creatures on one level should fall to the next

Creatures already fall (unless they can fly). This PR is about the rest.

Most of the stuff works as described: falls when unsupported. Internally I implemented it by adding a "support dirty" cache, which contains all tiles from which floor was recently removed.
Each turn the contents of the cache are checked for stuff that should fall, then the cache is emptied.

Big edit:
- Disabled vehicles falling for now. Due to cache issues and inherent 2D-ality of vehicle collision code, it is quite big and complex. I'll finish it in a separate PR.
- Items and furniture fall more or less correctly. "TINY" or "NOCOLLIDE" furniture can be crushed by dropping stuff on it.
- Furniture dropping on creatures, vehicles, furniture and terrain will hurt that, itself and contained items. Didn't balance it yet - you can kill hulks by dropping chairs at them. You can even punch through floor by dropping a safe on it.
- Falling items don't hurt yet, nor do they get damaged by the fall.
- I disabled "recursive" falling of stuff. One object can fall only once per turn for now. This simplifies things like furniture "waiting" for a free spot, sitting in the air until creature below it moves. I could re-add it later on if it's desired.

Balancing not included. It's experimental mode only.
I removed the mapgen hack I was talking about below. It doesn't cause much performance gain and it's a hack.

### CODE REVIEW

**Comment by DavidKeaton** [comment]
As for telling how full a tile is, what are your thoughts on using what percentage a square's volume is used?

**Comment by Coolthulhu** [comment]
The issue is less "how full" and more "is it full already?".

**Comment by kevingranade** [comment]
Vehicles landing on vehicles, apply ramming, calculating speed based on 1 meter of G acceleration per level fallen?  The same could apply for vehicles landing on furniture.
The solution for vehicles falling into water is that they fall if they don't have "enough" wheels on solid ground to support them.  This works fine for driving off a cliff, but not as well for driving over irregularly spaced holes in the ground.

**Comment by DavidKeaton** [comment]
Would you merely check length of vehicle, then if >50% of it is over, then cause it to fall? Curious.

Also, I gotcha Coolthulhu, that isn't the easiest to determine without explicit dimensions of objects.

**Comment by Coolthulhu** [comment]
But how to make a partially hanging vehicle fall?

The current solution for water simply despawns a vehicle that has more than 66% of wheels submerged. This isn't possible with a cliff.

The best solution I can think of right now would be to slightly push the vehicle in the direction of weighted average of hanging wheels, so that it eventually slides down the cliff fully.
But then some weird Dwarf Fortress shenanigans could be constructed to make auto-travelling vehicles sliding on impulse ramps.

**Comment by kevingranade** [comment]
One possibility is to use the center of mass for the vehicle, which we know and is not necessarily the geometric center of the vehicle, but the problem is that not every hole is going to be a sheer cliff.  One tricky example is a https://en.wikipedia.org/wiki/Bus_trap where all wheels might be on solid ground, but the center of mass is over a gap.

**Comment by kevingranade** [comment]
Yea, sliding into the pit is the best I could come up with also, but it doesn't solve te problem without some kind of way to detect that we're wedged.

**Comment by DavidKeaton** [comment]
well we have different tires, yeah? since there is only a handful, why not add a height to each? We could perhaps then say that a tire's height / 2 gives it the distance it can go up or become banked at?

**Comment by Coolthulhu** [comment]
Getting clogged in the mud/water/whatever is a much easier problem to solve.
The deal is: some tiles can be fallen through. To the lower z-level.
Wheel size won't help here, unless we're talking about wheels way larger than a human.

**Comment by Coolthulhu** [comment]
I ran across a small (design) problem:
I'd like to check for falling furniture on every non-mapgen `ter_set`, `furn_set` and `add_item_or_charges`.
Would it be acceptable to add a flag "is during mapgen" to `map` class? If not, what would be the best way to handle it without rewriting every line with `ter_set` invocation?

It didn't cause any noticeable performance loss during my testing because it's just adding points to a set and then checking them once each, but it may be just my PC. It's a toaster, but not a total toaster (2 ghz core duo).

**Comment by kevingranade** [comment]
I'm not sure I like that concept.  On one hand there's the handy tinymap/map distinction already, where tinymaps are mostly (only?) used for mapgen, but on the other hand I foresee that being an assumption that breaks weirdly in the future.  My inclination is to add the check to the existing functions, and perhaps push the actuall setting of terrain down into a helper function that does not do the extra checks, and then do a second pass and have parts of mapgen call the inner "dangerous" function as needed based on profiling.

**Comment by DavidKeaton** [comment]
why not store the address of two similar functions, one for mapgen one without, then after mapgen switch it to your one?

```
static std::array<std::function, 2> mapgen_func = {{&mapgen_ter_set, &coolthulhu_ter_set}};
mapgen_func[is_done_with_mapgen]();
```
If that makes sense.

**Comment by vache** [comment]
tinymap is used outside of mapgen.  Missions are one place I can think of.

**Comment by BevapDin** [comment]
Regarding the sealed container problem, the realistic solution is to add another item list to the map, so you have:
- items in the sealed container (accessible via special function, e.g. for the gas stations or the vending machines)
- items on the open ground, accessible like usual via `map::i_at` - this used normally (e.g. for field damage, for bashing damage, for monster interaction)
- (maybe) buried items

> Vehicle landing damage - how heavy, should it depend on horizontal speed

My basic physics knowledge says horizontal speed should not affect this at all.

> Furniture landing on stuff and damage. We don't have pianos to drop on people

But we need them. For the proper "zombie kill of the week" (-:

**Comment by Coolthulhu** [comment]
Updated the description.

This PR took me wayyy too much time. And was way more frustrating than it looks like. I limited its scope, due to how complex it turned out to be.

I'll follow up on it, but for now I think it is more or less complete.

File: src/map.cpp [hunk 1]
```diff
@@ -1775,6 +1822,253 @@ int map::climb_difficulty( const tripoint &p ) const
...
+    const auto check_tile = [this]( tripoint &pt ) {
+        tripoint dest( pt );
+        if( has_floor( dest ) ) {
+            pt = dest;
+            return SS_FLOOR;
+        }
+
+        tripoint below_dest( pt.x, pt.y, pt.z - 1 );
+        if( supports_above( below_dest ) ) {
+            pt = dest;
```
Comment (BevapDin) [inline src/map.cpp]: That assignment does not seem to do anything (same for the one above and at the end of the function).

`dests` is initialized to be a copy of `pt` and neither of those points is changed here. it seems like a no-op and should be removed.

**Comment by BevapDin** [comment]
Well, the items and furniture reacts a bit slowly, it takes a whole turn until it falls down.

Vehicles should probably ignore `tr_ledge` because it destroys the wheel and that is completely unrealistic. I can push the front wheel of my bike over a ledge and the wheel just hangs in midair, but it is *not* damaged at all.

Old bug: The remote vehicle control does not work across z-levels. Test case: street sweeper on z-level 0, player character on z-level 1. Getting to control the car works, but after only few turns this happens: "Lost contact with the vehicle." (note: it's not due to distance that would give a different message). Probably because "remote_controlling_vehicle" only stores a 2D point, see `game::remoteveh`.

**Comment by Coolthulhu** [comment]
`tr_ledge` was me unrolling the vehicle-related changes. Later on it will not ignore the ledge, but use it to check if it should start falling.

Remote vehicles may not work because vehicle cache on multiple z-levels has issues. Not sure how do they happen, but they also prevent falling vehicles from acting properly.

**Pull request merged**